apps/sync/js/adapters/history.js

```diff
@@ -0,0 +1,309 @@
+      });
+    });
+  }
+
+  function mergeRecordsToDataStore(existed, newPlace) {
+    if (existed.url !== newPlace.url) {
+      // The existed record has different url(id) with the new record.
+      throw new Error('Inconsistent records');
+    }
+    if (!existed.fxsyncId) {
+      existed.fxsyncId = newPlace.fxsyncId;
+    } else if(existed.fxsyncId !== newPlace.fxsyncId) {
+      // Two records have different fxSyncId but have the same url(id).
+      throw new Error('Inconsistent records');
+    }
+
+    existed.visits = existed.visits || [];
+    if (existed.visits.length === 0 && newPlace.title) {
+      existed.title = newPlace.title;
+    } else if (newPlace.visits[0] >= existed.visits[0]) {
+      existed.title = newPlace.title;
+    }
+
+    newPlace.visits.forEach(item => {
+      if (existed.visits.indexOf(item) === -1) {
+        existed.visits.push(item);
+      }
+    });
+
+    existed.visits.sort((a, b) => {
+      // sort in descending order
+      return b - a;
+    });
+
+    return existed;
+  }
+
+  function addPlace(place) {
+    // 1. get place by url(id of DataStore)
+    // 2.A merge the existed one and new one if it's an existed one,
+    //     and update the places.
+    // 2.B Add a new record with RevisionId.
+
+    var id = place.url;
+    var revisionId;
+    return _ensureStore().then(placesStore => {
+      revisionId = placesStore.revisionId;
+      return placesStore.get(id);
+    }).then(existedPlace => {
+      if (existedPlace) {
+        var newPlace = mergeRecordsToDataStore(existedPlace, place);
+        return placesStore.put(newPlace, id, revisionId);
+      } else {
+        return placesStore.add(place, id, revisionId);
+      }
+    }).catch(e => {
+      console.error(e);
+      return;
+    });
+  }
+
+  function updatePlaces(places) {
+    return new Promise(resolve => {
+      places.reduce((cur, next) => {
+        return cur.then(() => {
+          if (next.deleted) {
+            return deletePlaceByFxSyncId(next.fxsyncId);
+          } else {
+            return addPlace(next);
+          }
+        });
+      }, Promise.resolve()).then(resolve);
+    });
+  }
+
+  function traverseRecords(each, revisionId) {
+    return _ensureStore().then(placesStore => {
+      return new Promise(resolve => {
+        var cursor = placesStore.sync(revisionId);
+        runNextTask(cursor);
+
+        function runNextTask(cursor) {
+          cursor.next().then(function(task) {
+            manageTask(cursor, task);
+          });
+        }
+
+        function manageTask(cursor, task) {
+          if (task.operation === 'done') {
+            resolve(task);
+            return;
+          }
+          each(task).then(() => {
+            runNextTask(cursor);
+          });
+        }
+      });
+    });
+  }
+
+  function deletePlace(id) {
+    return _ensureStore().then(placesStore => {
+      return placesStore.remove(id);
+    });
+  }
+
+  function deletePlaceByFxSyncId(fxsyncId) {
```

Comment:
Having that many helpers (with no clear order) makes the code a little bit hard to follow. Rename this to `deletePlace`, please.